dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java

```diff
@@ -767,4 +1034,656 @@ private String createOrderType(int offset) {
+    for (String itemUid : cteContext.getCteKeys()) {
+      CteDefinition cteDef = cteContext.getDefinitionByItemUid(itemUid);
+
+      // Handle Program Stage CTE (potentially with multiple offsets)
+      if (cteDef.isProgramStage()) {
+        for (Integer offset : cteDef.getOffsets()) {
+          String alias = cteDef.getAlias(offset);
+          builder.leftJoin(
+              itemUid,
+              alias,
+              tableAlias ->
+                  tableAlias
+                      + ".enrollment = ax.enrollment AND "
+                      + tableAlias
+                      + ".rn = "
+                      + (offset + 1));
+        }
+      }
+
+      // Handle 'Exists' type CTE
+      if (cteDef.isExists()) {
+        builder.leftJoin(itemUid, "ee", tableAlias -> tableAlias + ".enrollment = ax.enrollment");
+      }
+
+      // Handle Program Indicator CTE
+      if (cteDef.isProgramIndicator()) {
+        String alias = cteDef.getAlias();
+        builder.leftJoin(itemUid, alias, tableAlias -> tableAlias + ".enrollment = ax.enrollment");
+      }
+
+      // Handle Filter CTE
+      if (cteDef.isFilter()) {
+        String alias = cteDef.getAlias();
+        builder.leftJoin(itemUid, alias, tableAlias -> tableAlias + ".enrollment = ax.enrollment");
+      }
+    }
+  }
+
+  /**
+   * Computes a zero-based offset for use with the SQL <em>row_number()</em> function in CTEs that
+   * partition and order events by date (e.g., most recent first).
+   *
+   * <p>In this context, an {@code offset} of 0 typically means “the most recent event” (row_number
+   * = 1), a positive offset means “the Nth future event after the most recent” (for example, offset
+   * = 1 means row_number = 2), and a negative offset means “the Nth older event before the most
+   * recent”.
+   *
+   * <p>Internally, this method transforms the supplied {@code offset} into a
+   * <strong>zero-based</strong> index, suitable for comparing against the row_number output. For
+   * instance:
+   *
+   * <ul>
+   *   <li>If {@code offset == 0}, returns {@code 0}.
+   *   <li>If {@code offset > 0}, returns {@code offset - 1} (i.e., offset 1 becomes 0-based 0).
+   *   <li>If {@code offset < 0}, returns the absolute value ({@code -offset}).
+   * </ul>
+   *
+   * @param offset an integer specifying how many positions away from the most recent event
+   *     (row_number = 1) you want to select. A positive offset selects a future row_number, a
+   *     negative offset selects a past row_number, and zero selects the most recent.
+   * @return an integer representing the zero-based offset to use in a {@code row_number} comparison
+   */
+  private int computeRowNumberOffset(int offset) {
+    if (offset == 0) {
+      return 0;
+    }
+
+    if (offset < 0) {
+      return (-1 * offset);
+    } else {
+      return (offset - 1);
+    }
+  }
+
+  private void generateFilterCTEs(EventQueryParams params, CteContext cteContext) {
```

Comment:
Nice to have javadoc here explaining a bit and providing some output example of a SQL, in very high-level.